extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -0,0 +1,397 @@
+   * @param {!UnsubscribeDef} handler
+   */
+  unsubscribe(handler) {
+    this.unsubscribe_.push(handler);
+  }
+
+  /**
+   * Adds the event handler that will be called when all visibility conditions
+   * have been met.
+   * @param {function()} handler
+   */
+  onEvent(handler) {
+    this.eventPromise_.then(handler);
+  }
+
+  /**
+   * Sets visibility of this object. See `getVisibility()` for the final
+   * visibility calculations.
+   * @param {number} visibility
+   */
+  setVisibility(visibility) {
+    this.ownVisibility_ = visibility;
+    this.update();
+  }
+
+  /**
+   * Sets whether this object is blocked. Blocking means that visibility is
+   * not ready to be calculated, e.g. because an element has not yet
+   * sufficiently rendered. See `getVisibility()` for the final
+   * visibility calculations.
+   * @param {boolean} blocked
+   */
+  setBlocked(blocked) {
+    this.blocked_ = blocked;
+    this.update();
+  }
+
+  /**
+   * Returns the final visibility. It depends on the following factors:
+   *  1. This object's visibility.
+   *  2. Whether the object is blocked.
+   *  3. The parent's visibility.
+   * @return {number}
+   */
+  getVisibility() {
+    const ownVisibility = this.ownVisibility_ * (this.blocked_ ? 0 : 1);
+    if (!this.parent_) {
+      return ownVisibility;
+    }
+    if (this.factorParent_) {
+      return ownVisibility * this.parent_.getVisibility();
+    }
+    return ownVisibility * (this.parent_.getVisibility() > 0 ? 1 : 0);
+  }
+
+  /**
+   * Runs the calculation cycle.
+   */
+  update() {
+    const visibility = this.getVisibility();
+    this.update_(visibility);
+    if (this.children_) {
+      for (let i = 0; i < this.children_.length; i++) {
+        this.children_[i].update();
+      }
+    }
+  }
+
+  /**
+   * Returns the calculated state of visibility.
+   * @param {time} startTime
+   * @return {!Object<string, string|number>}
+   */
+  getState(startTime) {
+    return {
+      // Observed times, relative to the `startTime`.
+      firstSeenTime: timeBase(this.firstSeenTime_, startTime),
+      lastSeenTime: timeBase(this.lastSeenTime_, startTime),
+      lastVisibleTime: timeBase(this.lastVisibleTime_, startTime),
+      fistVisibleTime: timeBase(this.fistVisibleTime_, startTime),
+
+      // Durations.
+      maxContinuousVisibleTime: this.maxContinuousVisibleTime_,
+      totalVisibleTime: this.totalVisibleTime_,
+
+      // Visibility percents.
+      loadTimeVisibility: this.loadTimeVisibility_ * 100 || 0,
+      minVisiblePercentage: this.minVisiblePercentage_ * 100,
+      maxVisiblePercentage: this.maxVisiblePercentage_ * 100,
+    };
+  }
+
+  /**
+   * @param {number} visibility
+   * @private
+   */
+  update_(visibility) {
+    if (!this.eventResolver_) {
+      return;
+    }
+    // Update state and check if all conditions are satisfied
+    const conditionsMet = this.updateCounters_(visibility);
+    if (conditionsMet) {
+      if (this.scheduledRunId_) {
```

Comment:
We can combine three cases by moving this above the `!this. eventResolver_` check: 

1. [Line 258](https://github.com/ampproject/amphtml/pull/7812/files#diff-b669887883c9dad04d04ef833c3c0001R258)
2. [Line 268](https://github.com/ampproject/amphtml/pull/7812/files#diff-b669887883c9dad04d04ef833c3c0001R268)
3. [Line 273](https://github.com/ampproject/amphtml/pull/7812/files#diff-b669887883c9dad04d04ef833c3c0001R273)